File: P1/ejercicio4/FilterChain.java

package ejercicio4;

import java.util.ArrayList;
import java.util.List;

public class FilterChain {
    private List<Filter> filters = new ArrayList<>();
    private AuthTarget target;

    public void addFilter(Filter filter) {
        filters.add(filter);
    }

    public void setTarget(AuthTarget target) {
        this.target = target;
    }

    public void execute(Message message) throws Exception {
        for (Filter filter : filters) {
            filter.execute(message);
        }

        if (target != null) {
            target.authenticate(message);
        }
    }
}

File: P2/email_filter_app/lib/filters/filter_chain.dart
import 'package:email_filter_app/credentials.dart';
import 'package:email_filter_app/auth_target.dart';
import 'package:email_filter_app/filters/filter.dart';

class FilterChain {
  final List<Filter> _filters = [];
  AuthTarget? _target;

  void addFilter(Filter filter) {
    _filters.add(filter);
  }

  void setTarget(AuthTarget target) {
    _target = target;
  }

  void execute(Credentials credentials) {
    for (final filter in _filters) {
      filter.execute(credentials);
    }
    _target?.authenticate(credentials);
  }
}
